modules/storage-api/src/testFixtures/java/org/apache/ignite/internal/storage/index/impl/TestSortedIndexStorage.java

```diff
@@ -118,21 +114,31 @@ public Cursor<IndexRow> scan(
         boolean includeLower = (flags & GREATER_OR_EQUAL) != 0;
         boolean includeUpper = (flags & LESS_OR_EQUAL) != 0;
 
-        Stream<Map.Entry<ByteBuffer, Set<RowId>>> data = index.entrySet().stream();
-
-        if (lowerBound != null) {
-            ToIntFunction<ByteBuffer> lowerCmp = boundComparator(lowerBound, includeLower ? 0 : -1);
-
-            data = data.dropWhile(e -> lowerCmp.applyAsInt(e.getKey()) < 0);
+        if (!includeLower && lowerBound != null) {
+            setEqualityFlag(lowerBound);
         }
 
-        if (upperBound != null) {
-            ToIntFunction<ByteBuffer> upperCmp = boundComparator(upperBound, includeUpper ? 0 : 1);
+        if (includeUpper && upperBound != null) {
+            setEqualityFlag(upperBound);
+        }
 
-            data = data.takeWhile(e -> upperCmp.applyAsInt(e.getKey()) <= 0);
+        SortedMap<ByteBuffer, Set<RowId>> data;
+
+        if (lowerBound == null && upperBound == null) {
+            data = index;
+        } else if (lowerBound == null) {
+            data = index.headMap(upperBound.byteBuffer());
+        } else if (upperBound == null) {
+            data = index.tailMap(lowerBound.byteBuffer());
+        } else {
+            try {
+                data = index.subMap(lowerBound.byteBuffer(), upperBound.byteBuffer());
+            } catch (IllegalArgumentException e) {
```

Comment:
This is new for me, how can it happen?

Reply:
This exception is thrown if the lowerBound is larger than the upperBound

Reply:
Ok, thank you!